pkg/strategy/dca2/recover.go

```diff
@@ -0,0 +1,268 @@
+		default:
+			return None, fmt.Errorf("there is unexpected status %s of order %s", order.Status, order)
+		}
+	}
+
+	if filledCnt > 0 && cancelledCnt == 0 {
+		activeOrderBook.Add(openOrders...)
+		orderStore.Add(openOrders...)
+		return OpenPositionOrderFilled, nil
+	}
+
+	if openedCnt > 0 && filledCnt > 0 && cancelledCnt > 0 {
+		return OpenPositionOrdersCancelling, nil
+	}
+
+	if openedCnt == 0 && filledCnt > 0 && cancelledCnt > 0 {
+		return OpenPositionOrdersCancelled, nil
+	}
+
+	return None, fmt.Errorf("unexpected order status combination")
+}
+
+func recoverPosition(ctx context.Context, position *types.Position, queryService queryAPI, currentRound Round) error {
+	if position == nil {
+		return nil
+	}
+
+	var positionOrders []types.Order
+	position.Reset()
+	if currentRound.TakeProfitOrder.OrderID != 0 {
+		if !types.IsActiveOrder(currentRound.TakeProfitOrder) {
+			return nil
+		}
+
+		positionOrders = append(positionOrders, currentRound.TakeProfitOrder)
+	}
+
+	for _, order := range currentRound.OpenPositionOrders {
+		// no executed quantity order, no need to get trades
+		if order.ExecutedQuantity.IsZero() {
+			continue
+		}
+
+		positionOrders = append(positionOrders, order)
+	}
+
+	for _, positionOrder := range positionOrders {
+		trades, err := queryService.QueryOrderTrades(ctx, types.OrderQuery{
+			Symbol:  position.Symbol,
+			OrderID: strconv.FormatUint(positionOrder.OrderID, 10),
+		})
+
+		if err != nil {
+			return fmt.Errorf("failed to get trades of order (%d)", positionOrder.OrderID)
+		}
+
+		position.AddTrades(trades)
+	}
+
+	return nil
+}
+
+func recoverBudget(currentRound Round) fixedpoint.Value {
+	if len(currentRound.OpenPositionOrders) == 0 {
+		return fixedpoint.Zero
+	}
+
+	total := fixedpoint.Zero
+	for _, order := range currentRound.OpenPositionOrders {
+		total = total.Add(order.Quantity.Mul(order.Price))
+	}
+
+	if currentRound.TakeProfitOrder.OrderID != 0 && currentRound.TakeProfitOrder.Status == types.OrderStatusFilled {
+		total = total.Add(currentRound.TakeProfitOrder.Quantity.Mul(currentRound.TakeProfitOrder.Price))
+		for _, order := range currentRound.OpenPositionOrders {
+			total = total.Sub(order.ExecutedQuantity.Mul(order.Price))
+		}
+	}
+
+	return total
+}
+
+func recoverStartTimeOfNextRound(ctx context.Context, currentRound Round, coolDownInterval types.Duration) time.Time {
+	if currentRound.TakeProfitOrder.OrderID != 0 && currentRound.TakeProfitOrder.Status == types.OrderStatusFilled {
+		return currentRound.TakeProfitOrder.UpdateTime.Time().Add(coolDownInterval.Duration())
+	}
+
+	return time.Time{}
+}
+
+type Round struct {
+	OpenPositionOrders []types.Order
+	TakeProfitOrder    types.Order
+}
+
+func getCurrentRoundOrders(short bool, openOrders, closedOrders []types.Order, groupID uint32) (Round, error) {
+	openPositionSide := types.SideTypeBuy
+	takeProfitSide := types.SideTypeSell
+
+	if short {
+		openPositionSide = types.SideTypeSell
+		takeProfitSide = types.SideTypeBuy
+	}
```

Comment:
This can be extracted as a function for reusing.

Reply:
I wanna remove Short...... because there are too many things need to be considered, I wanna remove it first.